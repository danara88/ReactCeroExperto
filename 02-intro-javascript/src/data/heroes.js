export const heroes = [
    {
        id: 1,
        name: 'Batman',
        owner: 'DC'
    },
    {
        id: 2,
        name: 'Spiderman',
        owner: 'Marvel'
    },
    {
        id: 3,
        name: 'Superman',
        owner: 'DC'
    },
    {
        id: 4,
        name: 'Flash',
        owner: 'DC'
    },
    {
        id: 5,
        name: 'Wolverine',
        owner: 'Marvel'
    },
];


const heroesExportDefect =  [
    {
        id: 1,
        name: 'Batman',
        owner: 'DC'
    },
    {
        id: 2,
        name: 'Spiderman',
        owner: 'Marvel'
    },
    {
        id: 3,
        name: 'Superman',
        owner: 'DC'
    },
    {
        id: 4,
        name: 'Flash',
        owner: 'DC'
    },
    {
        id: 5,
        name: 'Wolverine',
        owner: 'Marvel'
    },
];

// Estamos expoertando el arreglo heroesExportDefect por defecto
export default heroesExportDefect;