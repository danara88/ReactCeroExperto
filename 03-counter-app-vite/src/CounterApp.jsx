import { useState } from 'react';
import PropTypes from 'prop-types';
/**
 * NOTES:
 * Cuando hacemos un cambio en el componente, tenemos que avisar a React
 * de que hubo un nuevo cambio dentro del componente y ocupamos
 * que redibujes esos cambios (React lo hace muy rapido y eficientemente).
 *
 * HOOKS:
 * Fueron introducidos en la version 16.8
 * No son mas que funciones.
 * Los hooks empiezan con "use".
 */

/**
 * Si el functional component se vuelve a renderizar,
 * no se va a crear un nuevo espacio en memoria para handleClick
 */
// const handleClick = (event) => {
//   console.log(event);
// };

/**
 * @method CounterApp
 * @description Functional component CounterApp
 */
export const CounterApp = ({ value = 0 }) => {
  // Cuando se cambia el estado, el componente se vuelve a renderizar
  console.log('render');
  /**
   * El valor inicial es 0.
   * setCounter nos ayudara a cambiar el valor del counter
   */
  const [counter, setCounter] = useState(value);

  //   const handleClick = () => setCounter(counter + 1);
  const handleAdd = () => setCounter((c) => c + 1);
  const handleSubstract = () => setCounter(counter - 1);
  const handleReset = () => setCounter(value);

  //  Anteriormente React renderizaba 4 veces el componetne aqui.
  //   Pero ahora primero ejecutac cada una de las funciones, luego cambiar el estado
  //   y luego vuelve a renderizar todo.
  //
  //   const handleClick = () => {
  //     setCounter(counter + 1);
  //     setCounter(counter + 1);
  //     setCounter(counter + 1);
  //     setCounter(counter + 1);
  //   };

  return (
    <>
      <h1>CounterApp</h1>
      <h2>{counter}</h2>
      <button onClick={handleAdd}>+1</button>
      <button onClick={handleSubstract}>-1</button>
      <button onClick={handleReset}>Reset</button>
    </>
  );
};

CounterApp.propTypes = {
  value: PropTypes.number.isRequired,
};
