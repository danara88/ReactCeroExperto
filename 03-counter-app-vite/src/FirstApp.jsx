// Los propTypes nos ayuda a ponerle tipo a las properties
import PropTypes from 'prop-types';

/**
 * NOTA:
 * Si no hay alguna dependencia de el componente y de alguna variable,
 * poner las variables y funciones fuera del componente.
 */

/**
 * Podemos pasar variables al HTML mientras no sean objetos
 */
const newMessage = {
  message: 'Hello',
  title: 'Daniel',
};

/**
 * Si la function getName lo pongo dentro de FirstApp,
 * al momento de redibujar el componente, va a volver a especificar
 * en memoria el espacio de la funcion.
 */
const getName = (value) => value;

const FirstApp = ({ title, subTitle }) => {
  // if (!title) {
  //     throw new Error("Title no existe");
  // }

  return (
    <>
      <h1 data-testid="test-title">{title}</h1>

      <p>{subTitle}</p>
      <p>{subTitle}</p>
      <code>{JSON.stringify(newMessage)}</code>
      <div>First App</div>
      <div>{getName('Daniel Aranda')}</div>
      {/* <p>soy un subtitulo</p> */}
    </>
  );
};

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string.isRequired,
};

// FirstApp.defaultProps = {
//   title: 'No hay titulo',
//   subTitle: 0,
// };

export default FirstApp;
