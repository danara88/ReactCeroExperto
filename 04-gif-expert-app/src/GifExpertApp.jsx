import { useState } from 'react';
import { AddCategory, GifGrid } from './components';

export const GifExperApp = () => {
  const [categories, setCategories] = useState(['Programming']);

  const onAddCategory = (newCategory) => {
    // const newCategories = [...categories];
    // newCategories.push('Valorant');

    if (categories.includes(newCategory)) return;

    // setCategories(newCategories);
    setCategories((categories) => [newCategory, ...categories]);
  };

  return (
    <>
      <h1>GifExpertApp</h1>
      {/* Todos los eventos comienzan con "on" */}
      <AddCategory onNewCategory={onAddCategory} />
      {/* Nunca utilizar la i porque React utiliza el valor para saber cuando un elemento se elimino */}
      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
