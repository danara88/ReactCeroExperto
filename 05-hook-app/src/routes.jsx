import { createBrowserRouter, Navigate } from 'react-router-dom';
import { MainApp } from './09-useContext/MainApp';
import { AboutPage, HomePage, LoginPage } from './09-useContext/pages';

export const getRoutes = () =>
  createBrowserRouter([
    {
      path: '/',
      element: <MainApp />,
      children: [
        {
          path: '/',
          element: <HomePage />,
        },
        {
          path: 'home',
          element: <HomePage />,
        },
        {
          path: 'login',
          element: <LoginPage />,
        },
        {
          path: 'about',
          element: <AboutPage />,
        },
        {
          path: '/*',
          element: <Navigate to="/home" />,
        },
      ],
    },
  ]);
