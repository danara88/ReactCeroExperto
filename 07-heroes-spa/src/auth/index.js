/**
 * ==============================
 * Auth module exports
 * ==============================
 */

export * from './components';
export * from './helpers';
export * from './hooks';
export * from './pages';
export * from './types';
