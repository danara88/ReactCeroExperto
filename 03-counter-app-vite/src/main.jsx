import React from 'react';
import ReactDOM from 'react-dom/client';
// import App from './HelloWorldApp';
// import FirstApp from './FirstApp';
import { CounterApp } from './CounterApp';
import './styles.css'; // Global styles

/**
 * Reacty y ReactDOM son necesarios para renderizar mi app.
 * Usualmente todo en React empieza con un componente (punto de entrada)
 */

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    {/* <App />
    <FirstApp /> */}
    <CounterApp value={10} />
  </React.StrictMode>
);
