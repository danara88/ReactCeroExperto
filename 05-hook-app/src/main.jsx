import ReactDOM from 'react-dom/client';
import './index.css';
// import { FocusScreen } from './04-useRef/FocusScreen';
// import { App } from './App';
// import './08-useReducer/';

// import { CounterCustomHook } from './01-useState/CounterCustomHook';
// import { CounterApp } from './01-useState/CounterApp';
// import { SimpleForm } from './02-useEffect/SimpleForm';
// import { FormWithCustomHook } from './02-useEffect/FormWithCustomHook';
// import { MultipleCustomHook } from './03-examples/MultipleCustomHook';
// import { Layout } from './05-useLayoutEffect/Layout';
// import { Memorize } from './06-memos/Memorize';
// import { MemoHook } from './06-memos/MemoHook';
// import { CallbackHook } from './06-memos/CallbackHook';
// import { Padre } from './07-tarea-memo/Padre';
// import { TodoApp } from './08-useReducer/TodoApp';

import { RouterProvider } from 'react-router-dom';
import { getRoutes } from './routes';

const router = getRoutes();

ReactDOM.createRoot(document.getElementById('root')).render(
  // <React.StrictMode>
  //{/* <CounterApp /> */}
  //{/* <CounterCustomHook /> */}
  // <SimpleForm />
  // <FormWithCustomHook />
  // <MultipleCustomHook />
  // <FocusScreen />
  // <Layout />
  // <Memorize />
  // <MemoHook />
  // <CallbackHook />
  // <Padre />
  // <TodoApp />
  <RouterProvider router={router} />
  // </React.StrictMode>
);
