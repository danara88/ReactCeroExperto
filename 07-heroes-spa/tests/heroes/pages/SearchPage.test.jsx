import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes';

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUseNavigate,
}));

describe('SearchPage', () => {
  beforeEach(() => jest.clearAllMocks());

  test('should display component with default values', () => {
    // Arrange
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  test('should display batman data and the input with query string value', () => {
    // Arrange
    const heroName = 'batman';
    render(
      <MemoryRouter initialEntries={['/search?q=' + heroName]}>
        <SearchPage />
      </MemoryRouter>
    );
    const searchInput = screen.getByRole('textbox');
    const heroImage = screen.getByRole('img');
    const alert = screen.getByTestId('alert');
    // Assert
    expect(searchInput.value).toBe(heroName);
    expect(heroImage.src).toContain('/heroes/dc-batman.jpg');

    // Dos formas de evaluar
    // Primera opcion
    expect(alert.style._values).toEqual({ display: 'none' });

    // Segunda opcion
    expect(alert.style.display).toBe('none');
  });

  test('should display error if a heroe was not found', () => {
    // Arrange
    const heroName = 'NotExistingHero';
    render(
      <MemoryRouter initialEntries={['/search?q=' + heroName]}>
        <SearchPage />
      </MemoryRouter>
    );
    const errorAlert = screen.getByTestId('error-alert');
    // Assert
    expect(errorAlert.style.display).toBe('');
  });

  test('should navigate to new screen', () => {
    // Arrange
    const inputValue = 'superman';
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    );
    const form = screen.getByTestId('search-form');
    const searchInput = screen.getByRole('textbox');
    // Act
    fireEvent.change(searchInput, { target: { name: 'searchText', value: inputValue } });
    fireEvent.submit(form);
    // Assert
    expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
