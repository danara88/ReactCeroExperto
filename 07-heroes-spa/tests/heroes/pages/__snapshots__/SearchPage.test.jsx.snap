// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should display component with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        data-testid="search-form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results:
      </h4>
      <hr />
      <div
        class="alert alert-primary   animate__animated animate__fadeIn"
        data-testid="alert"
      >
        Search a hero
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        data-testid="error-alert"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
